core/platform/auth/firebase_auth_services.py

```diff
@@ -393,3 +381,107 @@ def associate_multi_auth_ids_with_user_ids(auth_id_user_id_pairs):
+    """
+    # NOTE: "app" is the term Firebase uses for the "entry point" to the
+    # Firebase SDK. Oppia only has one server, so it only needs to instantiate
+    # one app.
+    firebase_connection = firebase_admin.initialize_app(
+        options={'projectId': feconf.OPPIA_PROJECT_ID})
+    try:
+        yield
+    finally:
+        # NOTE: This is not dangerous. We are just deleting the resources used
+        # to form a connection to Firebase servers.
+        firebase_admin.delete_app(firebase_connection)
+
+
+def _get_session_cookie(request):
+    """Returns the session cookie authorizing the signed in user, if present.
+
+    Args:
+        request: webapp2.Request. The HTTP request to inspect.
+
+    Returns:
+        str|None. Value of the session cookie authorizing the signed in user, if
+        present, otherwise None.
+    """
+    return request.cookies.get(feconf.FIREBASE_SESSION_COOKIE_NAME)
+
+
+def _get_id_token(request):
+    """Returns the ID token authorizing a user, or None if missing.
+
+    Oppia uses the OAuth 2.0's Bearer authentication scheme to send ID Tokens.
+
+    Bearer authentication (a.k.a. token authentication) is an HTTP
+    authentication scheme based on "bearer tokens", an encrypted JWT generated
+    by a trusted identity provider in response to login requests.
+
+    The name "Bearer authentication" can be understood as: "give access to the
+    bearer of this token." These tokens _must_ be sent in the `Authorization`
+    header of HTTP requests, and _must_ have the format: `Bearer <token>`.
+
+    Learn more about:
+        HTTP authentication schemes:
+            https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication
+        OAuth 2.0 Bearer authentication scheme:
+            https://oauth.net/2/bearer-tokens/
+        OpenID Connect 1.0 ID Tokens:
+            https://openid.net/specs/openid-connect-core-1_0.html#IDToken
+
+    Args:
+        request: webapp2.Request. The HTTP request to inspect.
+
+    Returns:
+        str|None. The ID Token of the request, if present, otherwise None.
+    """
+    scheme, _, token = request.headers.get('Authorization', '').partition(' ')
+    return token if scheme == 'Bearer' else None
+
+
+def _get_auth_claims_from_session_cookie(cookie):
+    """Returns claims from the session cookie, or None if invalid.
+
+    Args:
+        cookie: str|None. The session cookie to extract claims from.
+
+    Returns:
+        AuthClaims|None. The claims from the session cookie, if available.
+        Otherwise returns None.
+    """
+    if cookie:
+        try:
+            with _firebase_admin_context():
+                return _create_auth_claims(
+                    firebase_auth.verify_session_cookie(cookie))
+        except EXPECTED_SESSION_COOKIE_EXCEPTIONS:
+            # NOTE: logging.exception appends the stack trace automatically.
+            logging.exception('User session has ended and must be renewed')
+            raise
```

Comment:
To clarify: is this an expected error? If so, why raise? I thought you would raise if it's not one of the exceptions that are expected to occur.

(Though if we don't raise, what should we do?)

Reply:
Whoops, you're right. Fixed. We just return None instead (because it isn't an error for a session cookie to be missing, it just means the request isn't authenticated).

Reply:
Maybe explain that as a code comment?

Reply:
Done.